src/Dscp/Educator/Validate.hs

```diff
@@ -2,41 +2,104 @@
 -- | Validation of private block
 
 module Dscp.Educator.Validate
-       ( validatePrivateBlk
+       ( BlockValidationFailure (..)
+       , SubmissionValidationFailure (..)
+       , validatePrivateBlk
        ) where
 
 import Universum
 
-import Serokell.Util.Verify (verResToMonadError, verifyGeneric)
+import Data.Text.Buildable (build)
 
 import Dscp.Core (Address (..), SignedSubmission (..), Submission (..),
-                  SubmissionWitness (..))
-import Dscp.Crypto (fromFoldable, getMerkleRoot, hash, verify)
-import Dscp.Educator.Block (PrivateBlock (..), PrivateBlockHeader (..), pbBody, pbbTxs)
+                  SubmissionSig, SubmissionWitness (..))
+import Dscp.Crypto (Hash, PublicKey, MerkleSignature, fromFoldable, getMerkleRoot, hash, verify)
+import Dscp.Educator.Block (PrivateBlock (..), PrivateBlockHeader (..), _pbBody, _pbbTxs)
 import Dscp.Educator.Serialise ()
 import Dscp.Educator.Txs (PrivateTx (..))
 
+import qualified Text.Show
+import qualified Data.Text.Buildable ()
+
+
+-- | Block validation failures
+data BlockValidationFailure
+    = -- | Header merkle tree root does not match body merkle tree root
+      MerkleSignatureMismatch
+        { bvfExpectedSig     :: !(MerkleSignature PrivateTx)
+        , bvfActualMerkleSig :: !(MerkleSignature PrivateTx) }
+      -- | Submission failure
+    | SubmissionInvalid
+        { bvfSubmissionFailure :: !SubmissionValidationFailure }
+    deriving (Eq)
+
+-- | Submission validation failures
+data SubmissionValidationFailure
+    = -- | Witness public key do not match submission address
+      SubmissionPublicKeyMismatch
+        { svfExpectedPubKey :: !(Hash PublicKey)
+        , svfActualPubKey   :: !(Hash PublicKey) }
+      -- | Witness signature do not match submission signature
+    | SubmissionSignatureMismatch
+        { svfSubmissionHash   :: !(Hash Submission)
+        , svfSubmissionSigKey :: !PublicKey
+        , svfSubmissionSig    :: !SubmissionSig }
+    deriving (Eq)
+
+instance Show BlockValidationFailure where
+    show = toString . pretty
+
+instance Buildable BlockValidationFailure where
+    build (MerkleSignatureMismatch {..}) =
+      "Merkle tree root signature mismatch. Expected " `mappend` show bvfExpectedSig
+      `mappend` " got " `mappend` show bvfActualMerkleSig
+    build (SubmissionInvalid x) = build x
+
+instance Buildable SubmissionValidationFailure where
+    build (SubmissionPublicKeyMismatch {..}) =
+      "Submission public key address mismatch. Expected " `mappend` show svfExpectedPubKey `mappend`
+      " got " `mappend` show svfActualPubKey
+    build (SubmissionSignatureMismatch {..}) =
+      "Submission signature mismatch. Submission data " `mappend` show svfSubmissionHash
+      `mappend` " with signature " `mappend` show svfSubmissionSig
+      `mappend` " does not correspond to public key " `mappend` show svfSubmissionSigKey
```

Comment:
You can `mconcat [..., ..., ...]` or use `[qc|foo {svfSubmissionHash} bar]` after importing `Text.InterpolatedString.Perl6` from `interpolatedstring-perl6` for brevity. Latter one does not require `show` here and there.